src/lib/core/overlay/disable-body-scroll.ts

```diff
@@ -0,0 +1,74 @@
+import {Injectable, Optional, SkipSelf} from '@angular/core';
+import {ViewportRuler} from './position/viewport-ruler';
+
+
+/**
+ * Utilitity that allows for toggling scrolling of the viewport on/off.
+ */
+@Injectable()
+export class DisableBodyScroll {
+  private _bodyStyles: string = '';
+  private _htmlStyles: string = '';
+  private _previousScrollPosition: number = 0;
+  private _isActive: boolean = false;
+
+  /** Whether scrolling is disabled. */
+  public get isActive(): boolean {
+    return this._isActive;
+  }
+
+  constructor(private _viewportRuler: ViewportRuler) { }
+
+  /**
+   * Disables scrolling if it hasn't been disabled already and if the body is scrollable.
+   */
+  activate(): void {
+    let body = document.body;
+    let bodyHeight = body.scrollHeight;
+    let viewportHeight = this._viewportRuler.getViewportRect().height;
+
+    if (!this.isActive && bodyHeight > viewportHeight) {
+      let html = document.documentElement;
+      let initialBodyWidth = body.clientWidth;
+
+      this._htmlStyles = html.style.cssText || '';
```

Comment:
Rather than caching the styles and setting them back, what do you think of just adding a removing a CSS class from the body where all of the rules are `!important`? It might make things simpler.